libmisc/btrfs.c

```diff
@@ -0,0 +1,94 @@
+#include <linux/btrfs_tree.h>
+#include <linux/magic.h>
+#include <sys/statfs.h>
+
+#include "prototypes.h"
+
+
+static int run_btrfs_subvolume_cmd(const char *subcmd, const char *arg1, const char *arg2)
+{
+	int status = 0;
+	const char *cmd = "/sbin/btrfs";
+	const char *argv[] = {
+		strrchr(cmd, '/'),
+		"subvolume",
+		subcmd,
+		arg1,
+		arg2,
+		NULL
+	};
+
+	if (argv[0] == NULL)
+		argv[0] = cmd;
+	else
+		argv[0] = argv[0] + 1;
+
+	if (access(cmd, X_OK)) {
+		return 1;
+	}
+
+	if (run_command(cmd, argv, NULL, &status))
+		return -1;
+	return status;
+}
+
+
+int btrfs_create_subvolume(const char *path)
+{
+	return run_btrfs_subvolume_cmd("create", path, NULL);
+}
+
+
+int btrfs_remove_subvolume(const char *path)
+{
+	return run_btrfs_subvolume_cmd("delete", "-C", path);
+}
+
+
+/* Adapted from btrfsprogs */
```

Comment:
btrfsprogs is GPL, shadow is bsd style.  So I'm afraid we need a more precise definitino of "Adapted from' :(

It's so trivial it's difficult to imagine a problem, but...

Reply:
(If it's just "inspired by", or "informed by", i.e. it showed how to use the constnats, then that's probably ok)

Reply:
Good catch. Serge is raising a good question. *Sigh* licenses.

Reply:
This part comes from the libbtrfsutil/subvolume.c which is actually LGPL, but that doesn't really help. I think the most tell-tale copy/paste is the comment itself. The rest is just basically 2 or 3 lines here. Line 74 and line 92.

I'm not sure there really is any alternative way of writing those 2 lines. But I can remove that comment which is copied. Read access is not relevant anyway.

Reply:
Hm, I'd like to hear what @stgraber thinks.

Reply:
It does look like something that's sufficiently trivial and pretty much has only a single way of getting implemented that copyright is unlikely to apply. If we want to be extra safe here, it should be reasonably easy to find another piece of code out there, under a suitable license, which implements the same logic.

Reply:
Thanks.  You're all right, there's basically no other way of doing it, so let's take it.